pkg/plugin/streaming.go

```diff
@@ -0,0 +1,174 @@
+			Status: backend.SubscribeStreamStatusNotFound,
+		}, fmt.Errorf("expected tail in channel path")
+	}
+
+	var qr humio.Query
+	if err := json.Unmarshal(req.Data, &qr); err != nil {
+		return nil, err
+	}
+
+	err := ValidateQuery(qr)
+	if err != nil {
+		return nil, err
+	}
+
+	h.streamsMu.RLock()
+	defer h.streamsMu.RUnlock()
+
+	cache, ok := h.Streams[req.Path]
+	if ok {
+		msg, err := backend.NewInitialData(cache.Bytes(data.IncludeAll))
+		return &backend.SubscribeStreamResponse{
+			Status:      backend.SubscribeStreamStatusOK,
+			InitialData: msg,
+		}, err
+	}
+
+	return &backend.SubscribeStreamResponse{
+		Status: backend.SubscribeStreamStatusOK,
+	}, nil
+}
+
+func (h *Handler) PublishStream(context.Context, *backend.PublishStreamRequest) (*backend.PublishStreamResponse, error) {
+	return &backend.PublishStreamResponse{
+		Status: backend.PublishStreamStatusPermissionDenied,
+	}, nil
+}
+
+func formatDuration(d time.Duration) string {
+	if d.Hours() >= 24*365 {
+		return fmt.Sprintf("%.0fy", d.Hours()/(24*365))
+	}
+	if d.Hours() >= 24 {
+		return fmt.Sprintf("%.0fd", d.Hours()/24)
+	}
+	if d.Hours() >= 1 {
+		return fmt.Sprintf("%.0fh", d.Hours())
+	}
+	if d.Minutes() >= 1 {
+		return fmt.Sprintf("%.0fm", d.Minutes())
+	}
+	return fmt.Sprintf("%.0fs", d.Seconds())
+}
+
+// calculate relative time for live query
+func getQueryStartTime(req *backend.RunStreamRequest) (string, error) {
+	type TimeRange struct {
+		From string `json:"from"`
+		To   string `json:"to"`
+	}
+
+	type reqData struct {
+		Repository string    `json:"repository"`
+		QueryType  string    `json:"queryType"`
+		Version    string    `json:"version"`
+		TimeRange  TimeRange `json:"timeRange"`
+	}
+
+	var queryTimeRange reqData
+
+	err := json.Unmarshal(req.Data, &queryTimeRange)
+	if err != nil {
+		log.DefaultLogger.Error("Error converting JSON to struct:", err)
+		return "", err
+	}
+
+	fromTime := queryTimeRange.TimeRange.From
+	toTime := queryTimeRange.TimeRange.To
+
+	from, err1 := strconv.ParseInt(fromTime, 10, 64)
+	to, err2 := strconv.ParseInt(toTime, 10, 64)
+
+	if err1 != nil || err2 != nil {
+		log.DefaultLogger.Error("Error converting strings to integers:", err1, err2)
+		return "", err
+	}
+
+	timeDifference := to - from
+	duration := time.Duration(timeDifference) * time.Millisecond
+	relativeTime := formatDuration(duration)
+
+	return relativeTime, nil
+}
+
+func (h *Handler) RunStream(ctx context.Context, req *backend.RunStreamRequest, sender *backend.StreamSender) error {
+	var qr humio.Query
+	if err := json.Unmarshal(req.Data, &qr); err != nil {
+		return err
+	}
+	err := ValidateQuery(qr)
+	if err != nil {
+		return err
+	}
+
+	var queryStartTime string
+	queryStartTime, err = getQueryStartTime(req)
+	if err != nil {
+		return err
+	}
+
+	qr.Start = queryStartTime
+	c := make(chan humio.StreamingResults)
+	prev := data.FrameJSONCache{}
+	done := make(chan any)
+
+	h.QueryRunner.RunChannel(ctx, qr, c, done)
+
+	for {
```

Comment:
I'd like to understand in a multi-tenant world, how does this logic guarantee that two different streams won't receive each others data? Is this based on the sender?

Reply:
This is a great question - the logic that keeps streams isolated is `req.Path` which uniquely identifies each stream and relates it to tenant-specific query details from `req.Data`. It uses the cache to store data from each strem separately and then locks it, so multiple actions don’t conflict.